README.md
# Project Document 
- It is just for study, please don't use it for business.


src/history.txt
SQRQ:(1985)(北京大学)AND SQR:(北京大学) AND LeiXing:( FMSQ)
SQRQ:(1986)(北京大学)AND SQR:(北京大学) AND LeiXing:( FMSQ)
SQRQ:(1987)(北京大学)AND SQR:(北京大学) AND LeiXing:( FMSQ)
SQRQ:(1988)(北京大学)AND SQR:(北京大学) AND LeiXing:( FMSQ)
SQRQ:(1989)(北京大学)AND SQR:(北京大学) AND LeiXing:( FMSQ)
SQRQ:(1990)(北京大学)AND SQR:(北京大学) AND LeiXing:( FMSQ)
SQRQ:(1991)(北京大学)AND SQR:(北京大学) AND LeiXing:( FMSQ)
SQRQ:(1992)(北京大学)AND SQR:(北京大学) AND LeiXing:( FMSQ)
SQRQ:(1985)(北京大学)AND SQR:(北京大学) AND LeiXing:( FMSQ)
SQRQ:(1985)(北京大学)AND SQR:(北京大学) AND LeiXing:( FMSQ)
SQRQ:(1986)(北京大学)AND SQR:(北京大学) AND LeiXing:( FMSQ)
SQRQ:(1988)(北京大学)AND SQR:(北京大学) AND LeiXing:( FMSQ)
SQRQ:(1990)(北京大学)AND SQR:(北京大学) AND LeiXing:( FMSQ)
SQRQ:(1992)(北京大学)AND SQR:(北京大学) AND LeiXing:( FMSQ)
SQRQ:(1993)(北京大学)AND SQR:(北京大学) AND LeiXing:( FMSQ)
SQRQ:(1994)(北京大学)AND SQR:(北京大学) AND LeiXing:( FMSQ)
SQRQ:(1995)(北京大学)AND SQR:(北京大学) AND LeiXing:( FMSQ)
SQRQ:(1996)(北京大学)AND SQR:(北京大学) AND LeiXing:( FMSQ)
SQRQ:(1997)(北京大学)AND SQR:(北京大学) AND LeiXing:( FMSQ)
SQRQ:(1998)(北京大学)AND SQR:(北京大学) AND LeiXing:( FMSQ)
SQRQ:(1999)(北京大学)AND SQR:(北京大学) AND LeiXing:( FMSQ)
SQRQ:(2000)(北京大学)AND SQR:(北京大学) AND LeiXing:( FMSQ)



src/test.py
data=['A01B','A01C','A01D','A01F','A01G','A01H','A01J','A01K','A01L','A01M','A01N','A01P','A21B','A21C','A21D','A22B','A22C','A23B','A23C','A23D','A23F','A23G','A23J','A23K','A23L','A23N','A23P','A24B','A24C','A24D','A24F','A41B','A41C','A41D','A41F','A41G','A41H','A42B','A42C','A43B','A43C','A43D','A44B','A44C','A45B','A45C','A45D','A45F','A46B','A46D','A47B','A47C','A47D','A47F','A47G','A47H','A47J','A47K','A47L','A61B','A61C','A61D','A61F','A61G','A61H','A61J','A61K','A61L','A61M','A61N','A61P','A61Q','A62B','A62C','A62D','A63B','A63C','A63D','A63F','A63G','A63H','A63J','A63K','A99Z','C08B','B01B','B01D','B01F','B01J','B01L','B02B','B02C','B03B','B03C','B03D','B04B','B04C','B05B','B05C','B05D','B06B','B07B','B07C','B08B','B09B','B09C','B21B','B21C','B21D','B21F','B21G','B21H','B21J','B21K','B21L','B22C','B22D','B22F','B23B','B23C','B23D','B23F','B23G','B23H','B23K','B23P','B23Q','B24B','B24C','B24D','B25B','B25C','B25D','B25F','B25G','B25H','B25J','B26B','B26D','B26F','B27B','B27C','B27D','B27F','B27G','B27H','B27J','B27K','B27L','B27M','B27N','B28B','B28C','B28D','B29B','B29C','B29D','B29K','B29L','B30B','B31B','B31C','B31D','B31F','B32B','B33Y','B41B','B41C','B41D','B41F','B41G','B41J','B41K','B41L','B41M','B41N','B42B','B42C','B42D','B42F','B43K','B43L','B43M','B44B','B44C','B44D','B44F','B60B','B60C','B60D','B60F','B60G','B60H','B60J','B60K','B60L','B60M','B60N','B60P','B60Q','B60R','B60S','B60T','B60V','B60W','B61B','B61C','B61D','B61F','B61G','B61H','B61J','B61K','B61L','B62B','B62C','B62D','B62H','B62J','B62K','B62L','B62M','B63B','B63C','B63G','B63H','B63J','B64B','B64C','B64D','B64F','B64G','B65B','B65C','B65D','B65F','B65G','B65H','B66B','B66C','B66D','B66F','B67B','B67C','B67D','B68B','B68C','B68F','B68G','B81B','B81C','B82B','B82Y','B99Z','C03C','C04B','C05F','C08J','C08L','C09K','C10B','C10G','C10L','C10M','C11B','C11D','C12F','C12P','C22B','C25D','D01B','D01C','D01F','D01G','D01H','D06L','D21B','D21C','D21F','D21H','E03F','E04F','F23G','F23J','G03C','G03G','G21F','H01B','H01J','H01M','C01B','C01C','C01D','C01F','C01G','C02F','C03B','C05B','C05C','C05D','C05G','C06B','C06C','C06D','C06F','C07B','C07C','C07D','C07F','C07G','C07H','C07J','C07K','C08C','C08F','C08G','C08H','C08K','C09B','C09C','C09D','C09F','C09G','C09H','C09J','C10C','C10F','C10H','C10J','C10K','C10N','C11C','C12C','C12G','C12H','C12J','C12L','C12M','C12N','C12Q','C12R','C13B','C13C','C13D','C13F','C13G','C13H','C13J','C13K','C14B','C14C','C21B','C21C','C21D','C22C','C22F','C23C','C23D','C23F','C23G','C25B','C25C','C25F','C30B','C40B','C99Z','G01N','G01R','G03F','G06F','G06K','G06T','G09F','D01D','D02G','D02H','D02J','D03C','D03D','D03J','D04B','D04C','D04D','D04G','D04H','D05B','D05C','D06B','D06C','D06F','D06H','D06J','D06M','D06N','D06P','D06Q','D07B','D21D','D21G','D21J','D99Z','D06G','E01B','E01C','E01D','E01F','E01H','E02B','E02C','E02D','E02F','E03B','E03C','E03D','E04B','E04C','E04D','E04G','E04H','E05B','E05C','E05D','E05F','E05G','E06B','E06C','E21B','E21C','E21D','E21F','E99Z','F01B','F01C','F01D','F01K','F01L','F01M','F01N','F01P','F02B','F02C','F02D','F02F','F02G','F02K','F02M','F02N','F02P','F03B','F03C','F03D','F03G','F03H','F04B','F04C','F04D','F04F','F15B','F15C','F15D','F16B','F16C','F16D','F16F','F16G','F16H','F16J','F16K','F16L','F16M','F16N','F16P','F16S','F16T','F17B','F17C','F17D','F21H','F21K','F21L','F21S','F21V','F21W','F21Y','F22B','F22D','F22G','F23B','F23C','F23D','F23H','F23K','F23L','F23M','F23Q','F23R','F24B','F24C','F24D','F24F','F24H','F24J','F25B','F25C','F25D','F25J','F26B','F27B','F27D','F28B','F28C','F28D','F28F','F28G','F41A','F41B','F41C','F41F','F41G','F41H','F41J','F42B','F42C','F42D','F99Z','G01D','H03J','H04L','G04B','H01R','F23N','G01B','G01C','G01F','G01G','G01H','G01J','G01K','G01L','G01M','G01P','G01Q','G01S','G01T','G01V','G01W','G02B','G02C','G02F','G03B','G03D','G03H','G04C','G04D','G04F','G04G','G04R','G05B','G05D','G05F','G05G','G06C','G06D','G06E','G06G','G06J','G06M','G06N','G06Q','G07B','G07C','G07D','G07F','G07G','G08B','G08C','G08G','G09B','G09C','G09D','G09G','G10B','G10C','G10D','G10F','G10G','G10H','G10K','G10L','G11B','G11C','G12B','G21B','G21C','G21D','G21G','G21H','G21J','G21K','G99Z','H01H','H01C','H01F','H01G','H01K','H01L','H01P','H01Q','H01S','H01T','H02B','H02G','H02H','H02J','H02K','H02M','H02N','H02P','H02S','H03B','H03C','H03D','H03F','H03G','H03H','H03K','H03L','H03M','H04B','H04H','H04J','H04K','H04M','H04N','H04Q','H04R','H04S','H04W','H05B','H05C','H05F','H05G','H05H','H05K','H99Z']
from bs4 import BeautifulSoup;
from urllib import request;
import time;
import random;
file1=open('text.txt','w');
condition1='权利转移';
condition2='许可备案';
need=['合同备案号','让与人','变更前权利人','变更后权利人','授权公告日','许可种类','备案日期'];
storename=['htbah','ryr','bgqqlr','bghqlr','sqggr','rkzl','barq'];
pat="";

def getNum(gen):
    for a in gen:
        return a;
    return;

def get(dom):
    time.sleep(1.2+random.uniform(0,1));
    global pat;
    par=dom.parent.parent;
    patNum=getNum(par.select('td')[2].strings);
    pos=patNum.find('.');
    patNum=patNum[0:pos];
    if pat==patNum:
        return;
    pat=patNum;
    result={};
    result['zlh']=pat;
    result['fmmc']=par.select('td a')[0].string;
    result['fmr']=par.select('td')[5].string;
    curUrl='http://www.soopat.com/Home/SipoLegal/'+pat;
    curresponse=request.urlopen(curUrl);
    curpage=curresponse.read();
    curpage=curpage.decode('utf-8');
    cursoup=BeautifulSoup(curpage,'html.parser');
    message=cursoup.select('tr td[colspan=3]');
    
    whole_message=[];
    for m in message:
        for mm in m.stripped_strings:
            whole_message.append(mm);
            if mm=='授权':
                flag=False;
                for cursqrq in m.parent.parent.stripped_strings:
                    if flag:
                        result['sqrq']=cursqrq;
                        break;
                    if cursqrq=='法律状态公告日':
                        flag=True;
                # print(m.parent.parent);
    for m in whole_message:
        pm=m.find(':');
        if pm!=-1:
            for i in range(0,len(need)):
                if need[i]==m[0:pm] and not storename[i] in result:
                    result[storename[i]]=m[pm+1:];
    print(str(result));
    file1.write(str(result));
    file1.write('\n');
    return;

for curdata in data:
    first='http://www.soopat.com/Home/Result?SearchWord=FLH%3A('+curdata+')&PatentIndex=0&View=7&Valid=2';
    response=request.urlopen(first);
    page=response.read();
    page=page.decode('utf-8');
    soup=BeautifulSoup(page,'html.parser');
    if len(soup.select('p.right b'))==0:
        print('sb yanzhengma');
        break;
    pagenum=int(int(soup.select('p.right b')[0].string)/30);
    print(pagenum);
    if pagenum>34:
        pagenum=33;
    # print(pagenum);
    time.sleep(3);
    for i in range(0,pagenum):
        time.sleep(2+random.uniform(0,2));
        first='http://www.soopat.com/Home/Result?SearchWord=FLH%3A('+curdata+')&PatentIndex='+str(i*30)+'&View=7&Valid=2';
        print(first);
        response=request.urlopen(first);
        page=response.read();
        page=page.decode('utf-8');
        soup=BeautifulSoup(page,'html.parser');
        item=soup.select('tbody tr.td_left');
        for k in item:
            check=k.select('td div');
            for j in check:
                if(j.string==condition1 or j.string==condition2):
                    get(j);
                    break;



src/index.py
data=[];
school=['北京大学', '北京工业大学', '北京科技大学', '北京理工大学', '北京师范大学', '北京邮电大学', '东北农业大学', '哈尔滨工业大学', '合肥工业大学', '华南师范大学', '暨南大学', '辽宁大学', '南京大学', '南京理工大学', '宁夏大学', '清华大学', '上海交通大学', '石河子大学', '四川大学', '太原理工大学', '天津医科大学', '同济大学', '西北大学', '西北工业大学', '西北农林科技大学', '西藏大学', '中国地质大学', '中国矿业大学 （徐州）', '中国政法大学', '安徽大学', '北京航空航天大学', '北京化工大学', '北京交通大学', '北京林业大学', '北京中医药大学', '大连海事大学', '大连理工大学', '电子科技大学', '东北大学', '东北林业大学', '东北师范大学', '东华大学', '东南大学', '福州大学', '复旦大学', '广西大学', '哈尔滨工程大学', '海南大学', '河北工业大学', '河海大学', '湖南大学', '湖南师范大学', '华北电力大学', '华东理工大学', '华东师范大学', '华南理工大学', '华中科技大学', '华中农业大学', '华中师范大学', '吉林大学', '江南大学', '兰州大学', '南昌大学', '南京航空航天大学', '南京农业大学', '南京师范大学', '南开大学', '内蒙古大学', '青海大学', '厦门大学', '山东大学', '陕西师范大学', '上海大学', '四川农业大学', '苏州大学', '天津大学', '武汉大学', '武汉理工大学', '西安电子科技大学', '西安交通大学', '西南大学', '西南大学', '西南交通大学', '新疆大学', '延边大学', '云南大学', '长安大学', '浙江大学', '郑州大学', '中国海洋大学', '中国科学技术大学', '中国矿业大学（北京）', '中国农业大学', '中国人民大学', '中国石油大学（北京）', '中国石油大学（华东）', '中国药科大学', '中南大学', '中山大学', '重庆大学', '贵州大学'];
from bs4 import BeautifulSoup;
from urllib import request
import time;
import datetime;
import random;
import urllib.parse;
now=datetime.datetime.now();
#生成关键词
for s in school:
    for i in range(1985,2018):
        curkey='SQRQ:('+str(i)+')('+s+')AND SQR:('+s+') AND LeiXing:( FMSQ)';
        data.append(curkey);
filename=str(now.microsecond)+'.txt';#
file1=open(filename,'w');
file2=open('history.txt');
a=file2.read();
history=a.split('\n');
#history里面是已经爬过的所有信息
file3=open('history.txt','w');
file4=open('max.txt','w');
file3.write(a);
file3.write('\n');
condition1='权利转移';
condition2='许可备案';
need=['合同备案号','让与人','变更前权利人','变更后权利人','授权公告日','许可种类','备案日期','受让人'];
storename=['htbah','ryr','bgqqlr','bghqlr','sqggr','rkzl','barq','srr'];
pat="";

def getNum(gen):
    i=0;
    for a in gen:
        if i==0:
            ccc=a;
            i=i+1;
        else:
            return ccc+a;
    return;
def get(dom,tt):
    time.sleep(1.2+random.uniform(0,1));
    global pat;
    par=dom.parent.parent;
    patNum=getNum(par.select('td')[2].strings);
    pos=patNum.find('.');
    patNum=patNum[0:pos];
    print(patNum);
    if pat==patNum:
        return;
    pat=patNum;
    result={};
    result['zlh']=pat;
    result['type']=tt;
    result['fmmc']=par.select('td a')[0].string;
    result['fmr']=par.select('td')[5].string;
    result['sqr']=par.select('td a font')[0].string;
    result['sqrq']=par.select('td font')[1].string;
    curUrl='http://www.soopat.com/Home/SipoLegal/'+pat;
    print(curUrl);
    curresponse=request.urlopen(curUrl);
    curpage=curresponse.read();
    curpage=curpage.decode('utf-8');
    cursoup=BeautifulSoup(curpage,'html.parser');
    message=cursoup.select('tr td[colspan=3]');
    whole_message=[];
    for m in message:
        for mm in m.stripped_strings:
            whole_message.append(mm);
            if mm=='授权':
                flag=False;
                for cursqrq in m.parent.parent.stripped_strings:
                    if flag:
                        result['sqrq']=cursqrq;
                        break;
                    if cursqrq=='法律状态公告日':
                        flag=True;
            if mm=='公开':
                flag=False;
                for curgkrq in m.parent.parent.stripped_strings:
                    if flag:
                        result['gkrq']=curgkrq;
                        break;
                    if curgkrq=='法律状态公告日':
                        flag=True;
            if mm=='专利申请权、专利权的转移':
                flag=False;
                for zyrq in m.parent.parent.stripped_strings:
                    if flag:
                        result['zyrq']=zyrq;
                        break;
                    if zyrq=='法律状态公告日':
                        flag=True;
    ff=False;
    for m in whole_message:
        pm=m.find(':');
        if pm!=-1:
            for i in range(0,len(need)):
                if need[i]==m[0:pm] and not storename[i] in result:
                    ff=True;
                    result[storename[i]]=m[pm+1:];
    if ff==False:
        for m in whole_message:
            pm=m.find(':');
            if pm!=-1:
                result['whole']=m;
                break;
    
    print(str(result));
    file1.write(str(result));
    file1.write('\n');
    return;

for curdata in data:
    if curdata in history:
        continue;
    print(curdata);
    first='http://www.soopat.com/Home/Result?SearchWord='+urllib.parse.quote(curdata)+'&PatentIndex=0&View=7';
    response=request.urlopen(first);
    page=response.read();
    page=page.decode('utf-8');
    soup=BeautifulSoup(page,'html.parser');
    if len(soup.select('p.right b'))==0:
        print('请打开链接',first);
        print('然后输入验证码');
        print('然后在终端中输入 \'python3.6 index.py\'即可');
        break;
    pagenum=int(int(soup.select('p.right b')[0].string)/30);
    file3.write(curdata);
    file3.write('\n');
    if pagenum>32:
        pagenum=32;
        file4.write(curdata);
        file4.write('\n');
    print(pagenum);
    time.sleep(3);
    for i in range(0,pagenum):
        time.sleep(2+random.uniform(0,2));
        first='http://www.soopat.com/Home/Result?SearchWord='+urllib.parse.quote(curdata)+'&PatentIndex='+str(i*30)+'&View=7';
        print(first);
        response=request.urlopen(first);
        page=response.read();
        page=page.decode('utf-8');
        soup=BeautifulSoup(page,'html.parser');
        item=soup.select('tbody tr.td_left');
        for k in item:
            check=k.select('td div');
            for j in check:
                if(j.string==condition1):
                    get(j,0);
                    break;
                if(j.string==condition2):
                    get(j,1);
                    break;

